### PR TITLE
feat(brillig): loops

### LINKED ISSUE #1554: Generate Brillig code to handle for loops
### Problem

Add code to handle the following function:

```
unconstrained fn loop(x : u32) -> u32 {
  let sum = 0;
  for i in 0..x {
    sum = sum + i
 }
 sum
}
```

### Happy Case

.

### Alternatives Considered

_No response_

### Additional Context

_No response_

### Would you like to submit a PR for this Issue?

No

### Support Needs

_No response_

### PR DESCRIPTION
**Pull request by sirasistant**
# Description

## Problem\*

Resolves #1554

## Summary\*

This PR sets out to implement loops in brillig.

### Example

Before:

```

```

After:

```

```

## Documentation

- [ ] This PR requires documentation updates when merged.

  

  - [ ] I will submit a noir-lang/docs PR.

  

  - [ ] I will request for and support Dev Rel's help in documenting this PR.

  
  

## Additional Context

# PR Checklist\*

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

File: crates/noirc_evaluator/src/brillig/brillig_gen.rs
Comment (ludamad) [inline crates/noirc_evaluator/src/brillig/brillig_gen.rs:51]: what did dbg! used to do? (generally not a huge fan of these acronym names, unless comments follow them everywhere :D)

File: crates/noirc_evaluator/src/brillig/memory.rs
Comment (ludamad) [inline crates/noirc_evaluator/src/brillig/memory.rs:2]: As for when we 'get smarter', my vote is only get smarter once we have concrete numbers from a public VM implementation

File: crates/noirc_evaluator/src/ssa_refactor/opt/flatten_cfg.rs
Comment (ludamad) [inline crates/noirc_evaluator/src/ssa_refactor/opt/flatten_cfg.rs:146]: Brillig doesn't want flattening?

Comment (ludamad) [reply crates/noirc_evaluator/src/ssa_refactor/opt/flatten_cfg.rs:146]: This could use a comment I think

**Comment by ludamad** [comment]
It's getting close to turing complete :D lgtm so far

File: crates/noirc_evaluator/src/brillig/brillig_gen.rs
Comment (ludamad) [reply crates/noirc_evaluator/src/brillig/brillig_gen.rs:51]: Doh. That would be more obvious if not next to dfg

File: crates/noirc_evaluator/src/ssa_refactor/opt/flatten_cfg.rs
Comment (sirasistant) [reply crates/noirc_evaluator/src/ssa_refactor/opt/flatten_cfg.rs:146]: Brillig is generated before these steps are ran

Comment (sirasistant) [reply crates/noirc_evaluator/src/ssa_refactor/opt/flatten_cfg.rs:146]: will add comments!

Comment (ludamad) [reply crates/noirc_evaluator/src/ssa_refactor/opt/flatten_cfg.rs:146]: Ideally this wouldn't be a possible value, but probably not practical to have a new IR for each pass :D

File: crates/noirc_frontend/src/hir/type_check/expr.rs
Comment (sirasistant) [inline crates/noirc_frontend/src/hir/type_check/expr.rs:156]: See https://github.com/noir-lang/noir/pull/1583

**Review by ludamad — APPROVED** [review]
LGTM but let's see if kev thinks it's butter

File: crates/noirc_evaluator/src/ssa_refactor/opt/constant_folding.rs
Comment (jfecher) [inline crates/noirc_evaluator/src/ssa_refactor/opt/constant_folding.rs:30]: I think this step was added because of the bug that was fixed in #1611. Lets merge that PR then we can remove this snippet

**Review by jfecher — CHANGES_REQUESTED** [review]

**Review by jfecher — APPROVED** [review]
Approving this now since I'll be out tomorrow.

Lets still remove some of the extra changes that are no longer needed in constant_folding.rs, flatten_cfg.rs, and type_check/expr.rs

File: crates/noirc_frontend/src/hir/resolution/resolver.rs
Comment (kevaundray) [inline crates/noirc_frontend/src/hir/resolution/resolver.rs:310]: just a cargo fmt diff

File: crates/noirc_frontend/src/hir/type_check/expr.rs
Comment (kevaundray) [reply crates/noirc_frontend/src/hir/type_check/expr.rs]: 1583 is in the merge queue, I've tentatively merged it into this branch

File: crates/noirc_evaluator/src/ssa_refactor/opt/flatten_cfg.rs
Comment (kevaundray) [inline crates/noirc_evaluator/src/ssa_refactor/opt/flatten_cfg.rs:187]: Adding an issue for this

Comment (kevaundray) [reply crates/noirc_evaluator/src/ssa_refactor/opt/flatten_cfg.rs:187]: #1613

File: crates/noirc_evaluator/src/brillig/brillig_gen.rs
Comment (kevaundray) [reply crates/noirc_evaluator/src/brillig/brillig_gen.rs:122]: This should not happen, I'll open an issue for this -- removing this check will just add needless instructions, though this behavior indicates a bug

Comment (kevaundray) [reply crates/noirc_evaluator/src/brillig/brillig_gen.rs:122]: #1614

**Pull request merged**
